### ISSUE TITLE
Setting ratings doesn't work

### ISSUE THREAD

**Issue by chrisnorman7** [simon-weber/gmusicapi#478]
Retrieving a song from my library (api.get_all_songs()[0]), adding the previously non-existant key 'rating' with the value of '5' and running it through api.change_song_metadata gives this:
gmusicapi.exceptions.CallFailure: BatchMutateTracks: BatchMutateTracks: The serv
er reported failure while changing the requested resource. If this wasn't caused
 by invalid arguments or server flakiness, please open an issue.
(requests kwargs: {'headers': {'Authorization': '<omitted>', 'Content-Type': 'ap
plication/json'}, 'params': {'hl': 'en_US', 'tier': 'aa', 'alt': 'json'}, 'url':
 'https://mclients.googleapis.com/sj/v2.4/trackbatch', 'data': '{"mutations": [{
"update": {"composer": "", "trackType": "8", "nid": "Tq3nokjwqu2lhjfk7nofjborkki
", "artist": "Alison Krauss", "clientId": "22f0fecb-9b3a-4294-ba42-80515b435103"
, "discNumber": 1, "title": "Broadway (Live At Louisville Palace, USA/2002)", "k
ind": "sj#track", "albumArtist": "Alison Krauss", "lastModifiedTimestamp": "1436
773231003101", "estimatedSize": "9501284", "albumId": "Bxpwudewsyszlszjpy3aewzwd
xy", "artistArtRef": [{"url": "http://lh3.googleusercontent.com/iyS59d2CejhkgICU
DijItBsNmeIE648CDkYGyvBqSKPPILbHc0CHj7JN5DwwezRzUvkEL0Ic", "aspectRatio": "2", "
kind": "sj#imageRef", "autogen": false}], "storeId": "Tq3nokjwqu2lhjfk7nofjborkk
i", "albumArtRef": [{"url": "http://lh6.ggpht.com/Uxc1ViVCKMQq8S17K5nU8nT9SEWlRz
PKglslNhH8_ppXrpDi3ivA8-KzLRiZHhv-aFLo1ura", "aspectRatio": "1", "kind": "sj#ima
geRef", "autogen": false}], "recentTimestamp": "1436773230995000", "durationMill
is": "237000", "id": "7257df4c-69d8-3af5-879e-617c6cb5d718", "playCount": 7, "ge
nre": "Pop", "artistId": ["Ail5ifsxrie3thzvn4oaqpcgtxu"], "deleted": false, "rat
ing": "5", "creationTimestamp": "1430819297952725", "album": "Alison Krauss & Un
ion Station Live", "explicitType": "2", "trackNumber": 10, "year": 2009}}]}', 'm
ethod': 'POST'})
(response was: '{\n "mutate_response": [\n  {\n   "id": "7257df4c-69d8-3af5-879e
-617c6cb5d718",\n   "response_code": "INVALID_REQUEST"\n  }\n ]\n}\n')

Am I using it correctly? I was hoping to just build a dictionary of id and rating, because I'm storing tracks locally in an sqlite database, so if this could work some way too I'd love to know about it.

Cheers, and thank you so much for all your hard work.

**Comment by simon-weber** [comment]
Huh, I'm not sure what's wrong. The example from my test suite is passing and looks identical to how you're describing the call: https://github.com/simon-weber/gmusicapi/blob/develop/gmusicapi/test/server_tests.py#L636.

Can you try with just the id and rating keys? ie `s = api.get_all_songs()[0]; change_song_metadata({'id': s['id'], 'rating': '5'})`? Maybe your track has unchangeable metadata that mine doesn't.

**Comment by thebigmunch** [comment]
For what it's worth, I tested on a number of tracks as well with no issue.

**Comment by chrisnorman7** [comment]
OK... So I tried with it some more...


It seems that Google unhelpfully doesn't present you with an id field 
the whole time.


nid and storeId seem to be the store id (they both start with t), and 
you can't change ratings with that key.


This seems to be true in all cases except when getting them from the 
library, but I imagine you'd know more about that than I.


I tried chopping the t off the start of storeId (which I'm sure I read 
somewhere worked), but no dice.


So in short, if the id doesn't begin with a t, you can set the ratings 
no worries. Otherwise you're stuck.


Is this intentional? I mean it's entirely possible I'm missing the point 
of setting ratings... I thought you could rate any song even if it 
wasn't in your library, but that could be wrong.


Sorry for rambling comment! :-)


Cheers,

**Comment by thebigmunch** [comment]
It's expected that a non-library track would not have a library id. I hadn't tested rating a store track, since you had said you were rating a library song. But rating store tracks works just fine for me as well. Is it just one store track that isn't working for you? If so, what's the ``nid``?

**Comment by simon-weber** [comment]
This might be related to https://github.com/simon-weber/gmusicapi/issues/182.

**Comment by chrisnorman7** [comment]
Sorry...


I was rating a library song, but I didn't realise the difference between 
the different IDs, so I was storing the store ID instead of the regular 
ID in my database.


Now I've gone back to the console, I have some code which fails:


import sys
from gmusicapi.clients import Mobileclient as m

m = m()

m.login(sys.argv[1], sys.argv[2], m.FROM_MAC_ADDRESS)

s = m.search('anal cunt')['song_hits'][0]['track']

m.change_song_metadata({'id': s['nid'], 'rating': '1'})

This song is most definitely not in my music library. It gives me:

Traceback (most recent call last):
   File 
"C:\Users\CHRISN~1\ENVIRO~1\gmp3\lib\site-packages\gmusicapi\protocol\shared.py", 
line 249, in perform
     cls.check_success(response, parsed_response)
   File 
"C:\Users\CHRISN~1\ENVIRO~1\gmp3\lib\site-packages\gmusicapi\protocol\mobileclient.py", 
line 640, in check_success
     cls.__name__)
gmusicapi.exceptions.CallFailure: BatchMutateTracks: The server reported 
failure while changing the requested resource. If this wasn't caused by 
invalid arguments or server flakiness, please open an issue.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
   File "test.py", line 10, in <module>
     m.change_song_metadata({'id': s['nid'], 'rating': '1'})
   File "<decorator-gen-73>", line 2, in change_song_metadata
   File 
"C:\Users\CHRISN~1\ENVIRO~1\gmp3\lib\site-packages\gmusicapi\utils\utils.py", 
line 619, in wrapper
     return function(*args, **kw)
   File "<decorator-gen-72>", line 2, in change_song_metadata
   File 
"C:\Users\CHRISN~1\ENVIRO~1\gmp3\lib\site-packages\gmusicapi\utils\utils.py", 
line 596, in wrapper
     return function(*args, **kw)
   File 
"C:\Users\CHRISN~1\ENVIRO~1\gmp3\lib\site-packages\gmusicapi\clients\mobileclient.py", 
line 214, in change_song_metadata
     self._make_call(mutate_call, mutations)
   File 
"C:\Users\CHRISN~1\ENVIRO~1\gmp3\lib\site-packages\gmusicapi\clients\shared.py", 
line 84, in _make_call
     return protocol.perform(self.session, self.validate, *args, **kwargs)
   File 
"C:\Users\CHRISN~1\ENVIRO~1\gmp3\lib\site-packages\gmusicapi\protocol\shared.py", 
line 263, in perform
     raise_from(CallFailure(err_msg, e.callname), e)
   File "<string>", line 2, in raise_from
gmusicapi.exceptions.CallFailure: BatchMutateTracks: BatchMutateTracks: 
The server reported failure while changing the requested resource. If 
this wasn't caused by invalid arguments or server flakiness, please open 
an issue.
(requests kwargs: {'params': {'hl': 'en_US', 'alt': 'json', 'tier': 
'aa'}, 'method': 'POST', 'headers': {'Authorization': '<omitted>', 
'Content-Type': 'application/json'}, 'url': 
'https://mclients.googleapis.com/sj/v2.4/trackbatch', 'data': 
'{"mutations": [{"update": {"rating": "1", "id": 
"Tzendd23uvxmnjnoavcj4ajhsdi"}}]}'})
(response was: '{\n "mutate_response": [\n  {\n   "id": 
"Tzendd23uvxmnjnoavcj4ajhsdi",\n   "response_code": "INVALID_REQUEST"\n  
}\n ]\n}\n')

**Comment by thebigmunch** [comment]
The ``id`` field is for library tracks only. Store tracks would need an ``nid`` and/or ``storeId`` (I haven't tested if only one is neededed). Store tracks actually also need the ``trackType`` field as I discovered in #480.

Instead of ``m.change_song_metadata({'id': s['nid'], 'rating': '1'})``, you would do:

```
s['rating'] = 1
m.change_song_metadata(s)
```

**Comment by chrisnorman7** [comment]
Hi mate,

I can't really do that because I store the stuff in a db, and I leave 
out stuff like kind (although that could change). Is there anywhere that 
documents exactly what data is needed?


Also, from my own observations I find that track['nid'] == 
track['storeId'] all the time.

**Comment by thebigmunch** [comment]
> Also, from my own observations I find that track['nid'] ==
> track['storeId'] all the time.

Yes, but I didn't test if you need one or both in the dict for ``change_song_metadata``. I didn't finish testing that. I got down to a dict with just ``kind``, ``trackType``, ``nid``, and ``storeId``. You can check from there.

> I can't really do that because I store the stuff in a db, and I leave
> out stuff like kind (although that could change). Is there anywhere that
> documents exactly what data is needed?

While it's technically possible to rate songs with less, the documentation says to pass a song dictionary, not a subset of one. The example given is just like the code from my previous comment.

**Comment by chrisnorman7** [comment]
Yes, you're completely right. My thing now works!


Thank you so much.


By the way, if you are still happy to have GMP mentioned in the README, 
any chance you could update it to GMP3? The link for it is:

https://github.com/chrisnorman7/gmp3.git

**Comment by thebigmunch** [comment]
Just make a pull request updating the link. I'm sure Simon would have no problem accepting it.

**Comment by chrisnorman7** [comment]
Ah, woops... Sorry, didn't pay attention to who was replying to the comment.


Yeah, I'll do that!


Cheers,
